### ISSUE TITLE
Address reuse of malloc cannot be detected

### ISSUE THREAD

**Issue by peterschrammel** [diffblue/cbmc#2117]
CBMC returns VERIFICATION SUCCESSFUL for the following program, although the assertion can fail. 

```
void main()
{
  int *x = malloc(sizeof(int));
  free(x);
  int *y = malloc(sizeof(int));
  if(x == y) {
    // should be reachable
    assert(0);
  }
  free(y);
}
```

CBMC's object encoding assumes that malloc never returns the same address.

**Comment by kroening** [comment]
The equality check `x == y` is undefined behaviour.

**Comment by remi-delmas-3000** [comment]
The pointer comparison should be flagged as UB, but `--pointer-checks` does not catch this at the moment.

**Comment by feliperodri** [comment]
Is there any documentation with a list of all UBs that `--pointer-checks` covers?

**Comment by tautschnig** [comment]
May I elevate this to a discussion of the (absence of) tracking of object lifetimes? After all, addresses could only be reused when an earlier object at the same address has been freed (or gone out of scope). The following example (making use of knowledge of encoding internals) shows the other side of this, i.e., asserting that an object is accessible _before_ its lifetime starts:
````C
int *p;

void foo()
{
  int local;
  __CPROVER_assert(p == &local, "prepared pointer");
}

int main()
{
  p = 2ULL << ((sizeof(int*) - 1) * 8);
  __CPROVER_assert(__CPROVER_r_ok(p, sizeof(int)), "non-existent object is readable");
  foo();
}
````
This succeeds with `cbmc --no-simplify` (the combination of constant propagation and the simplifier picks out the integer value of the pointer, and then deems it to be out-of-bounds of the NULL object).

**Comment by jimgrundy** [comment]
There is a lot here.
- Comparison of pointers with different provenance is UB, but that doesn't give us a pass. It appears as if we have chosen a particular interpretation that makes this program correct.
- Certainly it looks like pointer comparison should generate the check that the compared pointers have the same object. Though it further confirms my feeling that we should not consider CBMC results sound unless pointer-checks are enabled.
- But what if we were to cast those pointers to long before doing the comparison, would we still know that the object ID test was required?
I don't know that tracking object lifetimes would solve this would it (though it might solve other problems)? It seems like even if the object above was live at the time of the assertion we should still have a problem because we have manufactured a pointer in an illegal way.

**Comment by remi-delmas-3000** [comment]
we need to decouple the identity of an allocation from the address at where it lives

**Comment by jimgrundy** [comment]
There is a long-term fix we need to contemplate that involves a richer model of pointers and malloc such that malloc can reuse previous freed addresses. This addresses Remi's point that

> we need to decouple the identity of an allocation from the address at where it lives

But for now, there is a bandaid that we need to apply to limit the soundness bleed that leverage's Daniel's comment that

> The equality check x == y is undefined behaviour

Pointer comparisons (==, <, <=, >, >=, !=) are UB for pointers with different provenance. Pointer values with different object-bits in CBMC have different provenance, so we could implement a check on all pointer comparisons that the object bits of the compared pointers match. That would improve the situation here because with --pointer-checks enabled (which should become the default in the next major release) the program in this issue would fail verification (because the pointer check would fail).

This is not a complete solution. You might be able to get around this by casting the pointers to integers before comparing. You might get around this by comparing them with XOR rather than ==, but we would catch more errors and reduce the risk of unsoundness. This should be fairly easy to do, so we should do it soon.

**Comment by thomasspriggs** [comment]
> There is a long-term fix we need to contemplate that involves a richer model of pointers and malloc such that malloc can reuse previous freed addresses. This addresses Remi's point that
> 
> > we need to decouple the identity of an allocation from the address at where it lives
> 
> But for now, there is a bandaid that we need to apply to limit the soundness bleed that leverage's Daniel's comment that
> 
> > The equality check x == y is undefined behaviour
> 
> Pointer comparisons (==, <, <=, >, >=, !=) are UB for pointers with different provenance. Pointer values with different object-bits in CBMC have different provenance, so we could implement a check on all pointer comparisons that the object bits of the compared pointers match. That would improve the situation here because with --pointer-checks enabled (which should become the default in the next major release) the program in this issue would fail verification (because the pointer check would fail).
> 
> This is not a complete solution. You might be able to get around this by casting the pointers to integers before comparing. You might get around this by comparing them with XOR rather than ==, but we would catch more errors and reduce the risk of unsoundness. This should be fairly easy to do, so we should do it soon.

I am not entirely sure what you mean by provenance in this context. I don't see that comparing pointers which point to different objects (have different object bits) should be considered to be unsound. So if we took the original example and removed the `free` then I think the resulting program has no undefined behaviour -
```
void main()
{
  int *x = malloc(sizeof(int));
  int *y = malloc(sizeof(int));
  if(x == y) {
    // should be well defined to be unreachable.
    assert(0);
  }
}
```
Therefore I reason that the undefined behaviour is to do comparisons on pointers which point to objects which have been freed. I think the pointer checks could be made to include instrumenting with such a check if they don't already, even with the current pointer model.

My understanding is that freeing a pointer in our current model adds that pointer to the value set of `__CPROVER_deallocated`. Therefore the instrumented code (following the original example) would look something like -
```
void main()
{
  int *x = malloc(sizeof(int));
  free(x);
  int *y = malloc(sizeof(int));
  assert(__CPROVER_deallocated != x); // Instrumentation on x
  assert(__CPROVER_deallocated != y); // Instrumentation on y
  if(x == y) {
    // should be reachable
    // Is still unreachable in our analysis, but instrumentation assertion on x can be violated.
    assert(0);
  }
  free(y);
}
```
Please someone correct me if my understanding of the undefined behaviour or our pointer model is incorrect.

**Comment by kroening** [comment]
Yes, adding a UB check for `==` fixes this, and this is likely the best fix. Some existing uses of `==` may need to be replaced by `__CPROVER_equal`.

If it bugs you that `malloc` can't return a previously used address, simply make it return `__CPROVER_deallocated` nondeterministically. However, this would not add anything on top of the above.

**Comment by thomasspriggs** [comment]
So to account for the possibility of non-zero offsets the pointer check instrumentation needs to be -
```
void main()
{
  int *x = malloc(sizeof(int));
  free(x);
  int *y = malloc(sizeof(int));
  assert(!__CPROVER_same_object(__CPROVER_deallocated, x)); // Instrumentation on x
  assert(!__CPROVER_same_object(__CPROVER_deallocated, y)); // Instrumentation on y
  if(x == y) {
    // should be reachable
    // Is still unreachable in our analysis, but instrumentation assertion on x can be violated.
    assert(0);
  }
  free(y);
}
```

**Comment by jimgrundy** [comment]
> I am not entirely sure what you mean by provenance in this context. I don't see that comparing pointers which point to different objects (have different object bits) should be considered to be unsound.

Best explanation of provenance I've found is in this paper: https://dl.acm.org/doi/10.1145/3290380
You may only need to read as far as section 2.

Note that different object-bits imply different provenance.

The paper will explain how the compiler may statically assume that x and y in the original example have different values because they have different provenance and hence the compiler is entitled to optimize away this condition (and with certain levels of optimization GCC will do just that). But, equally, if it is not optimized away at run-time that comparison may return true because that is what malloc may do in practice. Both behaviors are allowed because comparing pointers with different provenance is UB.

So, a sufficient check in this case is that when comparing x and y we assert that they have the same object.

For the pointer comparison with relational operations (<, <=, ...) the standard is clear:

> When two pointers are compared, the result depends on the relative locations in the address space
of the objects pointed to. If two pointers to object types both point to the same object, or both point
one past the last element of the same array object, they compare equal. If the objects pointed to
are members of the same aggregate object, pointers to structure members declared later compare
greater than pointers to members declared earlier in the structure, and pointers to array elements
with larger subscript values compare greater than pointers to elements of the same array with lower
subscript values. All pointers to members of the same union object compare equal. If the expression
P points to an element of an array object and the expression Q points to the last element of the same
array object, the pointer expression Q+1 compares greater than P. *In all other cases, the behavior is
undefined.*

If CBMC pointers have different object bits then they would be different objects under this definition.

**Comment by tautschnig** [comment]
1) As noted in multiple comments, we should expand `goto_check_ct::pointer_rel_check` to cover equals/not-equals. Some care, however, will be required to account for a) comparisons against `NULL` and b) "Two pointers compare equal if and only if [...] one is a pointer to one past the end of one array object and the other is a pointer to the start of a different array object that happens to immediately follow the first array object in the address space." (C11, 6.5.9.6) -- this case would fail the `same_object` test.
2) Daniel's suggestion to have `malloc` return `__CPROVER_deallocated` (likely under some constraints, such as same-object-size) might be the quickest fix to get us closer to modelling a typical free-list allocator that will, at times, return the same address.
3) Longer term we might be able to re-use shadow memory to do a very different pointer encoding: we could leave the actual pointer variable as a sequence of arbitrary bits and use shadow memory to encode the object and offset, respectively. We would thereby decouple the identity and the address as suggested by Remi (our other attempts to change the pointer encoding also worked towards this, but did all encoding entirely in the back-end, for example by creating bit vectors three times the width of the original pointer so as to have the integer address alongside the object and offset bits).

**Comment by thomasspriggs** [comment]
> 2\. Daniel's suggestion to have `malloc` return `__CPROVER_deallocated` (likely under some constraints, such as same-object-size) might be the quickest fix to get us closer to modelling a typical free-list allocator that will, at times, return the same address.


I think that making malloc potentially return a previously deallocated pointer is likely to run into issues relating to object sizes. My understanding is that the currently used pointer encoding requires each object ID to have a fixed size for the entirety of the analysis. If the original call to malloc and the following call to malloc which returns the deallocated pointer have different sizes, then this violates the assumption/requirement made by the pointer encoding.

**Comment by tautschnig** [comment]
> > 2. Daniel's suggestion to have `malloc` return `__CPROVER_deallocated` (likely under some constraints, such as same-object-size) might be the quickest fix to get us closer to modelling a typical free-list allocator that will, at times, return the same address.
> 
> I think that making malloc potentially return a previously deallocated pointer is likely to run into issues relating to object sizes. My understanding is that the currently used pointer encoding requires each object ID to have a fixed size for the entirety of the analysis. If the original call to malloc and the following call to malloc which returns the deallocated pointer have different sizes, then this violates the assumption/requirement made by the pointer encoding.

Agreed, and hence my "likely under some constraints, such as same-object-size." This, in turn, of course makes this an incomplete solution.

**Comment by jimgrundy** [comment]
> As noted in multiple comments, we should expand goto_check_ct::pointer_rel_check to cover equals/not-equals. Some care, however, will be required to account for a) comparisons against NULL and b) "Two pointers compare equal if and only if [...] one is a pointer to one past the end of one array object and the other is a pointer to the start of a different array object that happens to immediately follow the first array object in the address space." (C11, 6.5.9.6) -- this case would fail the same_object test.

That's a fun twist, so if you increment a pointer p one past the end of an array and that happens to be the same as a pointer q that points to a start of an array and p and q have equal values then that isn't UB. (Is what I understood that text to mean).
But, in CBMC could one ever pull that off? If we are taking about adding in a UB test for pointer equality that just asserts that the two pointers (if not NULL) must have the same object ID then what would the exception look like? It would have to be the case that:
- The size of the first object was the maximum legal size of an object.
- Then when we increment one past the end of the first object the offset bits would roll over and the object ID bits would increment, and now it would appear that the two pointers have the same object ID, so the assertion wouldn't trigger.
- But, I presume some other pointer arithmetic UB assertion in CBMC would trigger in the above case.

I think pointer provenance in C is a finer grained concept than object-id in CBMC. Two arrays can have different provenance in the C standard sense while having the same object ID in CBMB. For example, if you malloc a struct with two arrays, pointers into those two arrays will have different provenance but the same object-id. I think the text quoted likely refers to a situation like this. Adding this object-id test to equality comparison wouldn't flag every case UB under the standard because object-id's are course, but everything that was flagged I think is safe to flag. And, it would put a stop to the issue that started this discussion. 

I still believe that we need to implement that UB check on pointer equality tests.

**Comment by thomasspriggs** [comment]
> There is a lot here.
> 
> * Comparison of pointers with different provenance is UB, but that doesn't give us a pass. It appears as if we have chosen a particular interpretation that makes this program correct.

Comparison of pointers which have different object IDs, but which have not been freed should be well defined in my opinion. Take for example a linked list. If each node in the list is allocated with a separate call to `malloc`, then each node struct in the list will have a different object ID. It is desirable for equality comparisons of pointers to different nodes to be well defined behaviour so that it is possible to iterate over sub ranges of the linked list.

> * Certainly it looks like pointer comparison should generate the check that the compared pointers have the same object. Though it further confirms my feeling that we should not consider CBMC results sound unless pointer-checks are enabled.
> * But what if we were to cast those pointers to long before doing the comparison, would we still know that the object ID test was required?

Pointer checks are not applied to comparisons on types which are not pointers. So `(long)(&a) == (long)(&b)` would not have the same checks as `&a == &b`. The object ID (and offset) are encoded into the result of the cast as long as there are enough bits available to do so.

If `sizeof(long)` is the same as `sizeof(void *)` then casting the long back to a pointer will reproduce the original pointer and the same checks as for the original pointers will be applied when comparing this reconstructed pointer against other pointers. I am not sure if you intended to invoke truncation of the pointer in this example or not. 

>   I don't know that tracking object lifetimes would solve this would it (though it might solve other problems)? It seems like even if the object above was live at the time of the assertion we should still have a problem because we have manufactured a pointer in an illegal way.

**Comment by jimgrundy** [comment]
@thomasspriggs has changed my mind. Thank you Thomas. The UB for pointer comparisons with different provenance should be only for <,<=,>,>=, but not for == and !=. I guess one could still contrive to give a UB for this one since one of the pointers has been freed, but it's not really the solution I hoped it was. I guess we do need to look at a way to recycle allocations.

**Comment by kroening** [comment]
@jimgrundy Let me invite you to change your mind again. If we ban the use of deallocated memory in `==` and `!=`, which real-world piece of code would be invalidated?

**Comment by jimgrundy** [comment]
I'm totally OK with banning deallocated memory in == and !=. I had originally thought we would work this thing by disallowing pointer comparisons between pointers with different object IDs (which we do for inequalities), but that does rule out real-world code as Thomas points out.

What's our mechanism for checking that we aren't comparing a deallocated region? Do we use the one tracked free-ed region and add that to the pointer checks? Is that sufficient? (we often run into limitations with that).

**Comment by thomasspriggs** [comment]
> What's our mechanism for checking that we aren't comparing a deallocated region? Do we use the one tracked free-ed region and add that to the pointer checks? Is that sufficient? (we often run into limitations with that).

In our model of the C standard library functions we non-deterministically assign or don't assign the global variable `__CPROVER_deallocated` to the value of the pointer which is being freed. See https://github.com/diffblue/cbmc/blob/3d8cbbbda33a66918b793fd0ed83e4f30b24d15c/src/ansi-c/library/stdlib.c#L640 and https://github.com/diffblue/cbmc/blob/3d8cbbbda33a66918b793fd0ed83e4f30b24d15c/src/ansi-c/library/stdlib.c#L313

This means that the set of values which `__CPROVER_deallocated` may be assigned to is the set of pointers which have been de-allocated. That means that we can write asserts of the form `__CPROVER_assert(!__CPROVER_same_object(ptr, __CPROVER_deallocated));` Such an assertion may be shown to be violated if `ptr` holds a value which points within any freed object.

The only limitation of this approach is that we cannot show that two different pointers point to two different freed objects _in the same trace_. So if we have two different pointers we can insert two assertions like -
```C
   __CPROVER_assert(!__CPROVER_same_object(ptr1, __CPROVER_deallocated));
   __CPROVER_assert(!__CPROVER_same_object(ptr2, __CPROVER_deallocated));
```
Then if both `ptr1` may point to deallocated memory and `ptr2` may point to a different piece of deallocated memory then both of these problems can be shown in separate traces.

**Comment by remi-delmas-3000** [comment]
Hi @thomasspriggs @TGWDB , here is a better explanation on how to use shadow memory to simulate full width pointers and address reuse https://github.com/remi-delmas-3000/shadow-map/edit/main/address_reuse.c
